File: src/main/java/dev/slimevr/vr/Device.java
package dev.slimevr.vr;

import dev.slimevr.vr.trackers.Tracker;
import io.eiren.util.collections.FastList;

import java.net.InetAddress;
import java.util.concurrent.atomic.AtomicInteger;


public class Device {

	public static final AtomicInteger nextLocalDeviceId = new AtomicInteger();

	private final int id;
	private String customName;
	private String firmwareVersion;
	private String manufacturer;

	private final FastList<Tracker> trackers = new FastList<>();

	public Device() {
		this.id = nextLocalDeviceId.incrementAndGet();
	}

	public int getId() {
		return id;
	}

	public void setCustomName(String customName) {
		this.customName = customName;
	}

	public String getManufacturer() {
		return this.manufacturer;
	}

	public void setManufacturer(String manufacturer) {
		this.manufacturer = manufacturer;
	}

	public String getDisplayName() {
		return null;
	}

	public String getFirmwareVersion() {
		return this.firmwareVersion;
	}

	public void setFirmwareVersion(String firmwareVersion) {
		this.firmwareVersion = firmwareVersion;
	}

	public String getCustomName() {
		return this.customName;
	}

	public InetAddress getIpAddress() {
		return null;
	}

	public FastList<Tracker> getTrackers() {
		return trackers;
	}
}

File: server/core/src/main/java/dev/slimevr/tracking/trackers/Device.kt
package dev.slimevr.tracking.trackers

import dev.slimevr.tracking.trackers.udp.BoardType
import dev.slimevr.tracking.trackers.udp.MCUType
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.atomic.AtomicInteger

open class Device(val magSupport: Boolean = false) {
	open val id: Int = nextLocalDeviceId.incrementAndGet()
	open var name: String? = null
	open var firmwareVersion: String? = null
	open var manufacturer: String? = null
	open val trackers: MutableMap<Int, Tracker> = ConcurrentHashMap()

	/**
	 * Implement toString() to return a string that uniquely identifies the board type
	 * SHOULDN'T RETURN NULL WHEN toString() IS CALLED
	 */
	open val boardType: Any = BoardType.UNKNOWN
	open val mcuType: MCUType = MCUType.UNKNOWN

	open val hardwareIdentifier: String = "Unknown"

	open var logMessages: MutableList<String> = mutableListOf()

	val isOpenVrDevice: Boolean
		get() = manufacturer == "OpenVR"

	init {
		CoroutineScope(Job()).launch {
			// Wait a little for device to get configured
			delay(1000)
		}
	}

	/**
	 * Enables or disables magnetometers in all the trackers of the device
	 * if `sensorId` null or in the specified tracker
	 * @param sensorId If null, every sensor will be modified
	 */
	open suspend fun setMag(state: Boolean, sensorId: Int = 255) {
		TODO("Not implemented because no mag support: $magSupport")
	}

	companion object {
		@JvmStatic
		protected val nextLocalDeviceId = AtomicInteger()
	}
}
